testing/test_checkpoint_pipeline_runner.py

```diff
@@ -0,0 +1,232 @@
+    ],
+    ["2", "3"])
+
+tape_checkpoint_saved_after_second_step_test_arguments = (
+    [
+        ("a", TransformCallbackStep(tape.callback, ["1"])),
+        ("b", TransformCallbackStep(tape.callback, ["2"])),
+        ("c", SomeCheckpointStep(
+            checkpoint_data_inputs=data_inputs,
+            checkpoint_expected_outputs=expected_outputs)
+         ),
+        ("d", TransformCallbackStep(tape.callback, ["3"]))
+    ],
+    ["3"])
+
+tape_checkpoint_saved_after_last_step_test_arguments = (
+    [
+        ("a", TransformCallbackStep(tape.callback, ["1"])),
+        ("b", TransformCallbackStep(tape.callback, ["2"])),
+        ("c", TransformCallbackStep(tape.callback, ["3"])),
+        ("d", SomeCheckpointStep(
+            checkpoint_data_inputs=data_inputs,
+            checkpoint_expected_outputs=expected_outputs)
+         ),
+    ],
+    [])
+
+tape_checkpoint_saved_inside_subpipeline_last_step = (
+    [
+        ("a", TransformCallbackStep(tape.callback, ["1"])),
+        Pipeline([
+            ("b", TransformCallbackStep(tape.callback, ["2"])),
+            ("d", SomeCheckpointStep(
+                checkpoint_data_inputs=data_inputs,
+                checkpoint_expected_outputs=expected_outputs)
+             ),
+        ]),
+        ("e", TransformCallbackStep(tape.callback, ["3"])),
+        ("f", TransformCallbackStep(tape.callback, ["4"])),
+    ],
+    ["3", "4"])
+
+tape_checkpoint_saved_inside_subpipeline_first_step = (
+    [
+        ("a", TransformCallbackStep(tape.callback, ["1"])),
+        Pipeline([
+            ("d", SomeCheckpointStep(
+                checkpoint_data_inputs=data_inputs,
+                checkpoint_expected_outputs=expected_outputs)
+             ),
+            ("b", TransformCallbackStep(tape.callback, ["2"])),
+        ]),
+        ("e", TransformCallbackStep(tape.callback, ["3"])),
+        ("f", TransformCallbackStep(tape.callback, ["4"])),
+    ],
+    ["2", "3", "4"])
+
+tape_checkpoint_saved_inside_subpipeline_step_in_the_middle = (
+    [
+        ("a", TransformCallbackStep(tape.callback, ["1"])),
+        Pipeline([
+            ("b", TransformCallbackStep(tape.callback, ["2"])),
+            ("d", SomeCheckpointStep(
+                checkpoint_data_inputs=data_inputs,
+                checkpoint_expected_outputs=expected_outputs)
+             ),
+            ("e", TransformCallbackStep(tape.callback, ["3"])),
+        ]),
+        ("f", TransformCallbackStep(tape.callback, ["4"])),
+    ],
+    ["3", "4"])
+
+tape_checkpoint_saved_inside_subpipeline_of_subpipeline = (
+    [
+        ("a", TransformCallbackStep(tape.callback, ["1"])),
+        Pipeline([
+            ("b", TransformCallbackStep(tape.callback, ["2"])),
+            Pipeline([
+                ("e", TransformCallbackStep(tape.callback, ["3"])),
+                ("d", SomeCheckpointStep(
+                    checkpoint_data_inputs=data_inputs,
+                    checkpoint_expected_outputs=expected_outputs)
```

Comment:
Nice test setup. It would be cool trying with many chained checkpoints (e.g.: another test with checkpoitns in a and b and e too and see if they are skipped perhaps?